main/src/library/Fixpoint3/Boxing.flix

```diff
@@ -0,0 +1,778 @@
+        let mapping = Equality.computeMapping(program, withProv);
+        let boxingInfo = initializeInternal(rc, program, mapping, withProv);
+        let (facts, newBoxingInfo) = initializeFacts(rc, program, boxingInfo, mapping, withProv);
+        (newBoxingInfo, facts, mapping)
+
+    ///
+    /// Returns `(facts, boxing)` where `facts` consists of the `Int64` representatives of
+    /// the facts in `program`, with respect to `boxing`. Note that the old `boxingInfo` is
+    /// invalid for the returned `facts`.
+    ///
+    /// This procedure will be heavily simplified by https://github.com/flix/flix/issues/11034
+    /// where most of the code is removed.
+    ///
+    def initializeFacts(rc1: Region[r], program: RamProgram, boxingInfo: Boxing[r], mapping: RamIdToPos, withProv: Bool): (EFacts, Boxing[r]) \ r = region rc2 {
+        let RamProgram.Program(_, facts, _, _) = program;
+        let factsAsVec = Map.toVector(facts);
+        let counters = initializeCounters(rc1, boxingInfo);
+        let newFacts = Array.empty(rc2, Vector.length(factsAsVec));
+        let funcsToBeRun = factsAsVec |> Vector.mapWithIndex(i -> match (relSym, factVec) -> () -> {
+            let mappedResult = mapAllFacts(rc1, relSym, factVec, boxingInfo, counters, mapping, withProv);
+            Array.put((relSym, mappedResult), i, newFacts)
+        });
+        spawnFuncs(funcsToBeRun);
+        let newBoxing = fixBoxingAfterInit(rc1, counters, boxingInfo);
+        (Array.toMap(newFacts), newBoxing)
+    }
+
+    ///
+    /// Returns a correct `Boxing` constructed from `counters` and `boxingInfo`.
+    ///
+    /// After the simple initialization `boxingInfo` is missing the map back from `Int64`
+    /// to `Boxed`. This simply creates the it from the old `boxingInfo`.
+    ///
+    def fixBoxingAfterInit(rc: Region[r], counters: Vector[AtomicCounter[r]], boxingInfo: Boxing[r]): Boxing[r] \ r = {
+        let (_, boxToRepVec, typeInfo, locks) = boxingInfo;
+        let newBoxedArr = counters |> Vector.map(c -> Array.empty(rc, getOrCrash(Int64.tryToInt32(AtomicCounter.get(c)))));
+        let funcsToBeRun = newBoxedArr |>
+            Vector.mapWithIndex(i -> writeTo -> () -> {
+                match getType(i, typeInfo) {
+                    case Types.TyObject =>
+                        let boxToRep = Vector.get(i, boxToRepVec);
+                        boxToRep |> BPlusTree.forEach(val -> pos -> Array.put(val, getOrCrash(Int64.tryToInt32(pos)), writeTo))
+                    case _ => ()
+                }
+            });
+        spawnFuncs(funcsToBeRun);
+        let newMutLists = newBoxedArr |> Vector.map(Array.toMutList(rc));
+        (newMutLists, boxToRepVec, typeInfo, locks)
+    }
+
+    ///
+    /// Spawns all functions in `funcs` and await their termination.
+    ///
+    def spawnFuncs(funcs: Vector[Unit -> Unit \ ef]): Unit \ ef = region rc2 {
+        unchecked_cast({
+            let _: Unit = spawn unchecked_cast(
```

Comment:
This function just spawns one thread and runs them all there right? Should be `forEach` and then `spawn` inside?